### PR TITLE
WIP: Feat/edit jadwal

### PR DESCRIPTION
**Pull request by jonathanfilbert**
Features Added:
**FRONTEND**
- Edit Jadwal page with jadwal-editing functionality
- when editing jadwal, if the total SKS is 0, then a user has an option to delete the particular jadwal
- in view jadwal page, an edit button is added to take the users directly to editing page

**BACKEND**
- PUT endpoint to edit a jadwal based on a user's id and a jadwal's id

### CODE REVIEW

File: frontend/src/containers/EditSchedule/utils/index.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+            for (let k = 0; k < schedule.schedule_items.length; k++) {
+                if (schedule.schedule_items[k].name === courses.courses[i].classes[j].name) {
+                    if (!added[courses.courses[i].classes[j].name]) {
+                        result.push({ ...courses.courses[i].classes[j], credit: courses.courses[i].credit, parentName: courses.courses[i].name, term: courses.courses[i].term, name: courses.courses[i].classes[j].name });
+                        added[courses.courses[i].classes[j].name] = 1;
+                    }
+                }
+            }
+        }
+    }
```
Comment (jonathanfilbert) [inline frontend/src/containers/EditSchedule/utils/index.js]: Desperately in need of refactoring. This kind of code will not get into production 😂 @leleonardo81 @laymonage

Comment (laymonage) [reply frontend/src/containers/EditSchedule/utils/index.js]: We have a feature that lets you define your own agenda to be added to the schedule. So, `schedule_items` may contain items that do not exist in the current `courses`.

I'd suggest that:

- We make a copy of `schedule_items` e.g. `savedSchedule`; then
- in addition to `push`ing the adapted schedule object to the `result` array, we should also remove the matching schedule from the `savedSchedule` array; then
- after finishing the whole triple nested loop thing, we iterate through the remaining items in the `savedSchedule` (these are items that don't exist in the current `courses`: they may be custom agendas or courses from the previous terms), adapt them, and push them to the `results` array.

With these modifications, we'll be able to retain the "lost" schedule items.

**Review by laymonage — CHANGES_REQUESTED** [review]

File: frontend/src/containers/EditSchedule/utils/index.js (hunk 1 above)
Comment (jonathanfilbert) [reply frontend/src/containers/EditSchedule/utils/index.js]: Ah yes, didn't really think about the custom agendas when implementing this. Thanks for the proposed solutions, will implement this asap.

File: frontend/src/containers/EditSchedule/index.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+const EditSchedule = ({ match }) => {
+    const dispatch = useDispatch();
+    const auth = useSelector(state => state.auth);
+    const { isMobile } = useSelector(state => state.appState)
+    const { scheduleId } = useParams();
+    const [courses, setCourses] = useState(null);
+    const [detailData, setDetailData] = useState(null);
+    const [isCoursesDetail, setCoursesDetail] = useState(null);
+    const schedules = useSelector(state => state.schedules);
```
Comment (laymonage) [inline frontend/src/containers/EditSchedule/index.js]: Unused variable `schedules`?

Comment (jonathanfilbert) [reply frontend/src/containers/EditSchedule/index.js]: Fixed.

**Pull request merged**